compiler/rustc_infer/src/infer/error_reporting/mod.rs

```diff
@@ -2068,7 +2068,8 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
                                 if let Some(code) =
                                     code.strip_prefix('"').and_then(|s| s.strip_suffix('"'))
                                 {
-                                    if code.chars().nth(1).is_none() {
+                                    // Make sure the string literal is exactly 1 char long.
+                                    if code.chars().nth(1).is_none() && !code.is_empty() {
```

Comment:
Maybe swap conditions? Checking for emptiness when you previously checked something that can be non-empty, this reads quite ok.

Reply:
(not the PR author)
Would there be any downside to having `code.chars().count() == 1` as a single condition?

Reply:
The main downside is count() needs to scan the entire string, potentially taking a while on longer string constants, while the above approach completes in constant time regardless of the length.

Reply:
It may not be a significant amount of time- there's a tradeoff here between clarity and efficiency. I'm open to suggestions as to which we feel is more important in this case

Reply:
```
let mut code_chars = code.chars();
if code_chars.next().is_some() && code_chars.next().is_none() {
```
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=42cd8a25462ee83b32464a07c23509e4

Reply:
I don't think we need to worry about performance here, this code only runs when we're producing an error, it isn't in the fast path. I think this code would be clearer if it were just `code.chars().count() == 1`.

Reply:
Good point on this being a low-traffic path. Updating for clarity.